lib/TPP/Dialect/Xsmm/XsmmUtils.cpp

```diff
@@ -287,6 +358,175 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
   return failure();
 }
 
+FailureOr<int64_t> getLeadingDim(Type type, size_t pos) {
+  // Not shaped type, the leading dimension is the single scalar.
+  auto memref = dyn_cast<MemRefType>(type);
+  if (!memref)
+    return 1;
+  // For 1d memref we cannot use the stride as leading dimension, but the
+  // leading dimension is the dimension itself.
+  if (memref.getRank() == 1)
+    return memref.getShape()[0];
+
+  SmallVector<int64_t> strides;
+  int64_t offset;
+  if (failed(getStridesAndOffset(memref, strides, offset)))
+    return failure();
+  // fail if the strides are non-constant
+  if (llvm::any_of(strides, [](int64_t stride) {
+        return stride == ShapedType::kDynamic;
+      }))
+    return failure();
+  return strides[pos];
+}
+
+bool isTwoDTransposeOp(vector::TransposeOp transposeOp) {
+  auto operandType = dyn_cast<VectorType>(transposeOp.getOperand().getType());
+  bool isVnni = vnni::utils::isInVnniLayout(operandType.getRank(), operandType);
+  if (isVnni || operandType.getRank() != 2 ||
+      dyn_cast<VectorType>(transposeOp.getResult().getType()).getRank() != 2 ||
+      (isa<scf::ForallOp>(transposeOp->getParentOp()) &&
```

Comment:
This is unstable, as the parent op could be `scf::parallel` or `scfp:for` or `func::func` and you'd have to multiply the checks here.

Reply:
We run scf:foralltoparallelloop conversion after this pass. As for the funcop, its taken care of in this check already.